standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+## Docker standards
+
+### General
+- all containers should be created using Docker
+- Linux containers are preferred
+- containers should not be run using root user
+- containers should be scanned regularly for vulnerabilities
+
+### Parent images
+We should create Dockerfiles that are extended from a minimal Defra created parent images.  This will allow us to benefit from improved security and more efficient builds as we will not have to repeat steps that are common to all Dockerfiles.
+
+Different parent images should be created in line with each framework's best practice.  Eg, Node.js, .Net Core, Ruby etc.
+
+Parent Dockerfiles should be hosted in GitHub and be deployed to a container registry via CI/CD.
+
+Example Node.js parent image:
+```
+ARG NODE_VERSION=10.18.0
+FROM node:$NODE_VERSION-alpine
+ARG NODE_VERSION
+ARG VERSION=1.0.0
+
+# We need a basic init process to handle signals and reap zombie processes, tini handles that
+RUN apk update && apk add --no-cache tini=0.18.0-r0
+ENTRYPOINT ["/sbin/tini", "--"]
+
+# Never run as root, use the node user from the alpine image
+USER node
+# Default workdir for all the containers that use this image
+WORKDIR /home/node
+
+ENV NODE_ENV production
+# Set global npm dependencies to be stored under the node user directory
+ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
+ENV PATH=$PATH:/home/node/.npm-global/bin
+LABEL uk.gov.defra.ffc-node-base.version=$VERSION \
+      uk.gov.defra.ffc-node-base.node-version=$NODE_VERSION
+```
+
+#### Node.js
+- extend a minimal Node base image - e.g. node:alpine
+- set a fixed major version of the base image
+- set the `NODE_ENV` environment variable to production in parent
+- set global NPM dependencies to be stored under the node user directory:
+  ```
+  ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
+  ENV PATH=$PATH:/home/node/.npm-global/bin
+  ```
+- run as the `node` user and not `root`
+- ensure the working directory is owned by the node user - e.g. /home/node
+- make the exposed port configurable via a build argument and set as an environment variable:
+  ```
+  ARG PORT=3000
+  ENV PORT ${PORT}
+  ```
+
+#### .Net Core
+- a parent image should exist for both the .Net Core SDK and the .Net Core Runtime
+- extend minimal .Net Core base image
+- set a fixed major version of the base image
+- set the `ASPNETCORE_ENVIRONMENT` environment variable to production in parent
+- run as the `www-data` user and not `root`
+- ensure the working directory is owned by the www-data user - e.g. /app
+- a remote debugger should be added to the SDK file to aid development and testing
+  ```
+  RUN apt-get update \
+  && apt-get install -y --no-install-recommends unzip \
+  && curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg
+  ```
+
+## Container scanning
```

Comment:
I'd suggest tightening up the wording in this standard - normally we can talk about "containers" and mean a number of different things, but in this context I think it would be helpful to more clearly distinguish between "containers", "container definition files" and "container images".

It might just be me that's confused, but I'd be happier at least!

Reply:
Sure, I'll add an explanation at the top of the page.

Dockerfile = set of instructions for building a docker image
Image = a constructed set of layered docker instructions
Container = a running instance of an image

Reply:
Updated.